src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -0,0 +1,220 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.FindSymbols.Finders;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text.Adornments;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.CustomProtocol
+{
+    internal class FindUsagesLSPContext : FindUsagesContext
+    {
+
+        private const int MaxResultsChunkSize = 32;
+
+        private int _id = 0;
+
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        private readonly Dictionary<DefinitionItem, int> _definitionToId =
+            new Dictionary<DefinitionItem, int>();
+
+        private readonly List<VSReferenceItem> _resultsChunk =
+            new List<VSReferenceItem>();
+
+        private readonly IProgress<object[]> _progress;
+        private readonly Document _document;
+        private readonly int _position;
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+
+        public override CancellationToken CancellationToken { get; }
+
+        public FindUsagesLSPContext(
+            IProgress<object[]> progress,
+            Document document,
+            int position,
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            CancellationToken cancellationToken)
+        {
+            _progress = progress;
+            _document = document;
+            _position = position;
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+
+            CancellationToken = cancellationToken;
+        }
+
+        public async override Task OnCompletedAsync()
+        {
+            var results = Array.Empty<VSReferenceItem>();
+            using (await _semaphore.DisposableWaitAsync(CancellationToken).ConfigureAwait(false))
+            {
+                if (_resultsChunk.IsEmpty())
+                {
+                    return;
+                }
+
+                results = _resultsChunk.ToArray();
+                _resultsChunk.Clear();
+            }
+
+            _progress.Report(results);
+        }
+
+        public async override Task OnDefinitionFoundAsync(DefinitionItem definition)
+        {
+            using (await _semaphore.DisposableWaitAsync(CancellationToken).ConfigureAwait(false))
+            {
+                if (!_definitionToId.ContainsKey(definition))
+                {
+                    // Assinging a new id to the definition
+                    _definitionToId.Add(definition, _id);
+
+                    // VSReferenceItem currently doesn't support the ClassifiedTextElement type for DefinitionText,
+                    // so for now we just pass in a string.
+                    // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1100009
```

Comment:
is this still true?  @dibarbet  i thought LSP added support for this.

Reply:
I verified with Miguel within the past week or two that there isn't currently support for ClassifiedTextElement. https://devdiv.visualstudio.com/DevDiv/_workitems/edit/918138 Here's the current tracking bug for overall classification.